### PR TITLE
Support index lifecycle management

### PR DESCRIPTION
**Pull request by fpytloun**
DESCRIPTION HERE

(check all that apply)
- [ ] tests added
- [ ] tests passing
- [ ] README updated (if needed)
- [ ] README Table of Contents updated (if needed)
- [x] History.md and `version` in gemspec are untouched
- [x] backward compatible
- [ ] feature works in `elasticsearch_dynamic` (not required but recommended)

### CODE REVIEW

File: lib/fluent/plugin/out_elasticsearch.rb [hunk 1]
```diff
@@ -197,28 +198,17 @@ def configure(conf)
...
-            else
-              template_install(@template_name, @template_file, @template_overwrite, @enable_ilm, @deflector_alias, @ilm_policy_id)
-            end
-            create_rollover_alias(@index_prefix, @rollover_index, @deflector_alias, @application_name, @index_date_pattern, @enable_ilm, @ilm_policy_id, @ilm_policy)
-          end
-        elsif @templates
-          retry_operate(@max_retry_putting_template, @fail_on_putting_template_retry_exceed) do
-            templates_hash_install(@templates, @template_overwrite)
-          end
+      if @template_name && @template_file
```
Comment (cosmo0920) [inline lib/fluent/plugin/out_elasticsearch.rb]: Looks reasonable.
But, could you use `!Fluent::Engine.dry_run_mode` for using network code?
Fluentd `#configure` is run twice as dry run mode and normal mode.
I don't want to change previous behavior.

Comment (fpytloun) [reply lib/fluent/plugin/out_elasticsearch.rb]: Will fix, I forgot this conditional when moving the code.

**Comment by fpytloun** [comment]
@cosmo0920 I am thinking about one thing - every bulk write will now check presence of template and index alias. I don't see any performance impact so far on my environment but anyway I am thinking about some way how to avoid this and do this operation only once for each alias. Is it possible to have some caching or something? Or do you have better idea?

**Comment by cosmo0920** [comment]
> Is it possible to have some caching or something? Or do you have better idea?

I think that elasticsearch_dynamic's `dynamic_config` like caching mechanism can handle this issue. But, too old caching template clause should be dropped due to memory consumption.
If this PR doesn't makes heavily ES cluster operation, we can delay to tackle it on later.

File: lib/fluent/plugin/out_elasticsearch.rb (hunk 1 above)
Comment (cosmo0920) [reply lib/fluent/plugin/out_elasticsearch.rb]: I've fixed test failures and this comment fix.
Could you confirm them?

**Comment by fpytloun** [comment]
> > Is it possible to have some caching or something? Or do you have better idea?
> 
> I think that elasticsearch_dynamic's `dynamic_config` like caching mechanism can handle this issue. But, too old caching template clause should be dropped due to memory consumption.
> If this PR doesn't make heavily ES cluster operation, we can delay to tackle it on later.

Ok, added caching of index aliases, tested and seems it's working in my env. Don't suppose it should hold a lot of data to cause some memory leak issue as only single entry is added for each index alias.

**Comment by cosmo0920** [comment]
> > > Is it possible to have some caching or something? Or do you have better idea?
> > 
> > 
> > I think that elasticsearch_dynamic's `dynamic_config` like caching mechanism can handle this issue. But, too old caching template clause should be dropped due to memory consumption.
> > If this PR doesn't make heavily ES cluster operation, we can delay to tackle it on later.
> 
> Ok, added caching of index aliases, tested and seems it's working in my env. Don't suppose it should hold a lot of data to cause some memory leak issue as only single entry is added for each index alias.

Cool. I've added two testcases for your change.
And this PR also will fix the https://github.com/uken/fluent-plugin-elasticsearch/issues/565 with minor change. I'll work for it after merging this PR.

**Comment by fpytloun** [comment]
Thank you! :-)

Just in case someone wants use this feature before it's merge into elasticsearch-ruby, here's example Dockerfile that I am using:

```
FROM fluent/fluentd:v1.7.3-1.0

USER root
RUN apk add --update --virtual .build-deps \
        sudo build-base ruby-dev ruby-rake ruby-bundler zlib-dev curl-dev git \
 && gem install \
        fluent-plugin-azure-storage-append-blob \
        fluent-plugin-concat \
        fluent-plugin-detect-exceptions \
        #fluent-plugin-elasticsearch \
        fluent-plugin-flatten-hash \
        fluent-plugin-multi-format-parser \
        fluent-plugin-prometheus \
        fluent-plugin-rename-key \
        fluent-plugin-rewrite-tag-filter \
        fluent-plugin-secure-forward \
        fluent-plugin-throttle \
        aws-sdk-sqs \
        yajl-ruby \
 # XXX: Install patched elasticsearch-ruby
 && git clone -b support-ilm https://github.com/cosmo0920/elasticsearch-ruby.git /usr/src/elasticsearch-ruby \
 && cd /usr/src/elasticsearch-ruby && bundle install && find . -name "*.gemspec"|cut -d / -f 2 | while read i; do cd $i && cp LICENSE.txt LICENSE && rake install; cd ..; done \
 # XXX: Install patched elasticsearch plugin to support ILM
 && git clone -b support-index-lifecycle-management https://github.com/uken/fluent-plugin-elasticsearch.git /usr/src/fluent-plugin-elasticsearch \
 && cd /usr/src/fluent-plugin-elasticsearch && rake install \
 && cd / && rm -rf /usr/src/* \
 && gem sources --clear-all \
 && apk del .build-deps \
 && rm -rf /var/cache/apk/* \
           /home/fluent/.gem/ruby/2.3.0/cache/*.gem

ADD plugins /fluentd/plugins
#USER fluent

EXPOSE 24224

COPY entrypoint.sh /bin/
ENTRYPOINT ["tini",  "--", "/bin/entrypoint.sh"]
CMD ["fluentd"]
```

**Pull request merged**